DietPi-Software | Tor Relay

nonnorm opened the pull request:
<!--
Before submitting a pull request:
- Please ensure the target branch is "dev" (active development): https://github.com/MichaIng/DietPi/tree/dev
- Please ensure changes have been tested and verified functional.
-->
**Status**: Ready
- [x] Add IPv6 support
- [x] Add Unbound DNS resolution for exit relay
- [x] Add TOR Relay stats to `dietpi-cloudshell`
- [x] Test on VM
- [x] Test on RPi

**Reference**: https://github.com/MichaIng/DietPi/issues/3895

**Commit list/description**:
<!--
- DietPi-Config | Add "Fan control" option to "Performance Options"
![Screenshot](https://xxx.github.com/images/xxx.png)
- DietPi-Config | Add "Fan control" support for Odroid C2
- DietPi-Config | Syntax fix
-->
- DietPi-Software | Begin to add TOR relay
- DietPi-Software | CodeFactor Fixes
- DietPi-Software | Fixes
- DietPi-Software | Remove invalid text for optional fields
- DietPi-Software | Fix newlines in MENU_ARRAY
- DietPi-Software | Make sure that TOR relay and TOR hotspot don't conflict
- DietPi-Software | Remove spaces around equals signs
- DietPi-Software | Remove unneeded while loop
- DietPi-Software | Make sure nickname/email exists before adding it
- DietPi-Software | Add ControlPort
- DietPi-CloudShell | Add to dietpi-cloudshell
- DietPi-Software | Fix syntax
- DietPi-Software | Avoid conflict with BitWarden_RS
- DietPi-Software | Add Unbound DNS
- DietPi-Software | Add IPv6 Support
- DietPi-Software | Fix unbound resolution
- DietPi-CloudShell | Remove unneeded line
- DietPi-Software | Fix capitalization
- DietPi-Software | Fix syntax
- DietPi-Software | Add $
- DietPi-Software | Fixes
- DietPi-Software | Rename cancel button
- DietPi-Software | Simplify descriptions
- DietPi-Software | Simplify even more
- DietPi-Software | Clear invalid text
- DietPi-Software | Fix ipv6 setting
- DietPi-CloudShell | Increase MAX_SCENES
- DietPi-CloudShell | Add timeout to nyx
- DietPi-Software | Fix syntax
- DietPi-CloudShell | Rename nyxrc
- DietPi-CloudShell | Get nyx to stop and start
- DietPi-CloudShell | Add missing then
- DietPi-CloudShell | Try again to get nyx to work
- DietPi-Software | Fix nameserver change
- DietPi-CloudShell | Fix nyx scenes
- DietPi-CloudShell | Add Tor relay to disabled scenes
- DietPi-CloudShell | Make nyx update faster
- DietPi-Software | Fix IPv6
- DietPi-CloudShell | Fix nyx refreshing
- DietPi-CloudShell | Set NYX_ACTIVE variable
- DietPi-CloudShell | Kill nyx on exit
- DietPi-CloudShell | Only kil nyx if it is active
- DietPi-CloudShell | Change exit trap to only kill nyx if it is running
- DietPi-CloudShell | Switch to !=
- DietPi-CloudShell | Add !
- DietPi-CloudShell | Switch to ||
- DietPi-CloudShell | Fix NYX_ACTIVE placement
- DietPi-CloudShell | Fix syntax

nonnorm committed:
Begin to add TOR relay

nonnorm committed:
CodeFactor Fixes

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8568,6 +8585,222 @@ _EOF_
+
+				if G_WHIP_INPUTBOX "$(invalid_text)Pick a nickname for your relay. (This is optional)"; then
+
+					nickname=G_WHIP_RETURNED_VALUE
```
`$` missing here as you want to call the variable.
```
nickname=$G_WHIP_RETURNED_VALUE
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8568,6 +8585,222 @@ _EOF_
+
+				if G_WHIP_INPUTBOX "$(invalid_text)Pick a contact email for your relay. (This is optional but encouraged, and it will be published)"; then
+
+					email=G_WHIP_RETURNED_VALUE
```
Same as above
```
email=$G_WHIP_RETURNED_VALUE
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8568,6 +8585,222 @@ _EOF_
+
+					G_WHIP_MENU "${choice_required}Please choose which type of relay this should be set up as:" && break
+					choice_required='[ERROR] A choice is required to finish the TOR
+					 install.\n\n'
```
The two above lines should have been a single line, I guess?

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8568,6 +8585,222 @@ _EOF_
+_EOF_
+
+			G_CONFIG_INJECT 'NoNewPrivileges=' 'NoNewPrivileges=no' /lib/systemd/system/tor@default.service
+			G_CONFIG_INJECT 'NoNewPrivileges=' 'NoNewPrivileges=no' /lib/systemd/system/tor@.service
```
I'm fans of not editing the packages shipped service files directly, but instead place drop-in configs:
```
G_EXEC mkdir -p /etc/systemd/system/tor@.service.d
G_EXEC eval "echo -e '[Service]\nNoNewPrivileges=no' > /etc/systemd/system/tor@.service.d/dietpi-relay.conf"
```
I'm just not sure if this works for instantiated services, requires a quick test. If it works, `tor@default.service` does not need to be changed as it inherits everything from `tor@.service`.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8568,6 +8585,222 @@ _EOF_
+_EOF_
+
+			else
+				${aSOFTWARE_INSTALL_STATE[$software_id]}=-1
```
`$` is used for variable calls while here you want to assign a variable. Also, since you want to run an uninstall loop after this install loop, there needs to be another flag set:
```
aSOFTWARE_INSTALL_STATE[$software_id]=-1
UNINSTALL_REQUIRED=1
```

But actually, since the choice has been made before, I'm not sure if a separate YESNO makes sense here. Probably it is better to add the above warning/explanation directly into the `G_WHIP_MENU` choice loop? Either the menu text could contain further explanations (and the Exit relay warning) or, when Exit relay has been chosen, you add the dedicated YESNO directly into the same while loop so that when "NO" is chosen, users land in the parent G_WHIP_MENU to choose another relay type. This prevents the probably unexpected behaviour, when accidentally or intentionally choosing "NO", the whole already done install is uninstalled automatically.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1862,6 +1862,15 @@ DietPi-Software will decrypt and use it for software installs. You can change it
+		aSOFTWARE_DESC[$software_id]='add a node to the TOR network'
+		aSOFTWARE_TYPE[$software_id]=0
+		aSOFTWARE_CATEGORY_INDEX[$software_id]=17
+		aSOFTWARE_ONLINEDOC_URL[$software_id]='p=0#0'
```
Add:
```
aSOFTWARE_REQUIRES_USERINPUT[$software_id]=1
```
This assures that it cannot be installed via non-interactive firstrun setup which would cause an infinite menu loop 😉.

MichaIng commented:
Awesome, many thanks for all your implementation 👍 ❤️.

We need to prevent Tor Hotspot and Tor Relay from being installed concurrently. Easy when one has been installed already, but how to behave when both has been selected the same time, which one to select and which one to deselect, or deselect both to let user decide?

Would be of course great if we could make both running concurrently, but this would mean a crafted config file with both, relay and client sections, as it is not possible to use `/etc/tor/torrc.d/some.conf`, right?
__________
**EDIT: Ah I see your open task about this topic now. If we find a way, that would be nonplusultra, but otherwise it's okay to leave it either-or for now.**

nonnorm commented:
Maybe we could add another feature to the list, something like `aSOFTWARE_CONFLICTS_WITH=(1 2 3 4 etc.)` to handle this if I can't get them both to work concurrently?  Or I could just add it to the list of Pre-req software without `aSOFTWARE_REQUIRES` like Docker does.

nonnorm committed:
Fixes

nonnorm committed:
Remove invalid text for optional fields

nonnorm committed:
Fix newlines in MENU_ARRAY

MichaIng commented:
I have a larger rework in mind for dependencies and conflicts, exactly as you suggested with an array, which makes the individual `aSOFTWARE_REQUIRES_*` and other parts in the pre-reqs function obsolete. Until then I think it is best to add a check to the end of this pre-req function that sets the selected one to "0" if the other one was "2" and sets both to "0" if both have been selected concurrently, with meaningful error message. So any existing Tor install stays functional and if both are selected, none is installed to leave the choice to user.

nonnorm committed:
Make sure that TOR relay and TOR hotspot don't conflict

nonnorm committed:
Remove spaces around equals signs

nonnorm committed:
Remove unneeded while loop

nonnorm commented:
I'm trying to get the optional fields to work correctly, but it seems that `G_WHIP_INPUTBOX` already has a built in while loop and error handling.  For starters, I think this makes WireGuard's implementation redundant.  Also, the way that WireGuard implements it can lead to multiple error messages being shown.  Do you think that we should remove the loop from the default input box, or some other solution to make optional fields work?

nonnorm committed:
Make sure nickname/email exists before adding it

nonnorm committed:
Add ControlPort

nonnorm commented:
For `dietpi-cloudshell`, the TOR project has created a monitor called [`nyx`](https://nyx.torproject.org).  I think that it is the best candidate for `dietpi-cloudshell`, without writing our own Python application to connect with TOR.  It is fairly lightweight, except for the fact that it needs Python and the `stem` package.

nonnorm committed:
Add to dietpi-cloudshell

nonnorm committed:
Fix syntax

Joulinar commented on dietpi/dietpi-software:
```diff
@@ -1862,6 +1862,16 @@ DietPi-Software will decrypt and use it for software installs. You can change it
 		aSOFTWARE_ONLINEDOC_URL[$software_id]='p=221#p221'
 		aSOFTWARE_REQUIRES_BUILDESSENTIAL[$software_id]=1
+		#------------------
+		software_id=183
```
Software ID would need to be set to 184 to avoid conflicts with Bitwarden_rs

nonnorm committed:
Avoid conflict with BitWarden_RS

nonnorm committed:
Add Unbound DNS

Joulinar commented on dietpi/dietpi-software:
```diff
@@ -8722,6 +8722,8 @@ SocksPort 0
 ControlSocket 0
 _EOF_
 		elif [[ $G_WHIP_RETURNED_VALUE == 'Exit' ]]; then
+			G_AGI unbound
```
Would it make more sense to call `dietpi-software` installation for `unbound` ?

nonnorm replied:
While that probably would be preferred, I'm not sure how to do that from the configuration step.

MichaIng replied:
`dietpi-software` must not be called from within `dietpi-software`, so if the unbound install step is nothing else than the APT package install and the unbound config step comes below the Tor relay config step, it is fine as it is.

But if I am right, the unbound config step currently comes before these, around line 8400. So currently, installing the package here potentially breaks existing Pi-hole installs and other DNS servers running at port 53. And does Tor use it (127.0.0.1:53) automatically when it is there, since I don't see it being configured currently?

Joulinar replied:
What I meant was to use the already existing unbound installation and not to fire it up using `apt`. What about changing the order and move `TOR Relay` before `unbound` inside `dietpi-software`

nonnorm replied:
You're right, it does need to be configured.  TOR uses the DNS server being used by the system.

nonnorm committed:
Add IPv6 Support

nonnorm committed:
Fix unbound resolution

nonnorm committed:
Remove unneeded line

MichaIng commented:
Btw, a minor thing but I personally give this some weight: "Tor" is written "Tor" and not "TOR" 😉. Let's respect the official spelling choice: https://www.torproject.org/

MichaIng commented on dietpi/dietpi-services:
```diff
@@ -216,7 +216,8 @@ Available services:
 			aSERVICE_NAME+=('amiberry')			# DietPi Amiberry service
-      		aSERVICE_NAME+=('unbound')  # Unbound DNS server, could be currently used by machine
+			aSERVICE_NAME+=('unbound')  # Unbound DNS server, could be currently used by machine
+			aSERVICE_NAME+=('tor')  # TOR relay should not go down normally
```
What is the consequence when Tor goes down? Of course it breaks client access, but it does not affect the servers network access in any case, does it?

Generally, server services (used by clients that are not essential part of the server itself (network, remote control), are intended to be shut down during installs and maintenance tasks to minimise interference and resource usage. E.g. the just implemented Bitwarden_RS compiling and temporary Rust install require a lot of RAM and fail if that is not available. We applied some steps (check and in case increase swap size, limit concurrent compiling jobs), but if a high number of additional services is still running, that might not be enough anymore.

So as long as there is no forcing reason to keep Tor running, I suggest to let it usually handled by `dietpi-services`.

nonnorm committed:
Fix capitalization

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -2378,6 +2388,19 @@ DietPi-Software will decrypt and use it for software installs. You can change it
 		fi
 
+		# Don't let TOR Relay and TOR hotspot be installed at the same time
+		if (( ${aSOFTWARE_INSTALL_STATE[184]} < 0 && ${aSOFTWARE_INSTALL_STATE[61]} < 0 )); then
```
I think this should have been `> 0` 😄.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8385,6 +8416,231 @@ _EOF_
+			fi
+
+			local orport=443
+			invalid_text=
```
Make this a local variable:
```
local invalid_text=
```

nonnorm committed:
Fix syntax

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8385,6 +8416,231 @@ _EOF_
+			fi
+
+			echo "ControlPort 9051" >> /etc/tor/torrc
+			if [[ ! $nickname == '' ]]; then
```
Simpler check:
```
[[ $nickname ]]
```
has the meaning "is declared and not empty".

One liner:
```
[[ $nickname ]] && echo "Nickname $nickname" >> /etc/tor/torrc
```

Same below.

nonnorm committed:
Add $

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8385,6 +8416,231 @@ _EOF_
+				echo "ORPort [$ipv6]:$orport" >> /etc/tor/torrc
+			fi
+
+			G_EXEC systemctl restart tor
```
When Tor is controlled by `dietpi-services` (you agree to my suggestion to do so), then this restart should be removed.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -13053,6 +13309,14 @@ _EOF_
+		software_id=184 # TOR Relay
+		if (( ${aSOFTWARE_INSTALL_STATE[$software_id]} == -1 )); then
+
+			Banner_Uninstalling
```
Let's assure the service is removed cleanly:
```
			if [[ -f '/lib/systemd/system/tor.service' ]]
			then
				systemctl unmask tor
				systemctl disable --now tor
				rm -R /lib/systemd/system/tor.service*
			fi
			[[ -d '/etc/systemd/system/tor.service.d' ]] && rm -R /etc/systemd/system/tor.service.d
```

MichaIng commented:
Legend to have a DietPi-CloudShell scene for this, many thanks! 😃

MichaIng commented on dietpi/dietpi-cloudshell:
```diff
@@ -1492,6 +1498,18 @@ _EOF_
 			aEnabledScenes[$i]=1
+			# Configure Tor Relay
+			if [[ $i == 9 ]]; then
+				G_AGI nyx
```
Use `G_AG_CHECK_INSTALL_PREREQ nyx` here, which does not always call `apt-get` but only if the package is not yet installed. So after having this selected the first time, it will be done much quicker.

Also `G_AG_CHECK_INSTALL_PREREQ`  implies `apt-get update` which could otherwise be out-of-date. So outside of `dietpi-software` and `dietpi-update` that is basically the better replacement.

nonnorm commented:
There is still the problem with optional boxes. `G_WHIP_INPUTBOX` seems to have an input required, so even without a `while` loop, you can't leave the box blank.  Correct me if I'm wrong, but doesn't this also make having a `while` loop around the usage in `dietpi-software` redundant?
#
Okay, so you can skip boxes without a while loop by pressing escape/cancelling, but not by just leaving them blank.

nonnorm committed:
Fixes

MichaIng commented:
You could name the "Cancel" button "Skip" to make this clear. I think for required inputs, we added the while loop so that cancelling is not possible. By times I'll add this all to the `G_WHIP_INPUTBOX` function itself, allowing to force an input (although it must be assured that it is an interactive call) or leave an empty input or define a regex that must match the input (e.g. for port numbers, IP addresses, if no white spaces are allowed and such things).

nonnorm committed:
Rename cancel button

nonnorm committed:
Simplify descriptions

nonnorm committed:
Simplify even more

nonnorm committed:
Clear invalid text

nonnorm committed:
Fix ipv6 setting

nonnorm committed:
Increase MAX_SCENES

nonnorm committed:
Add timeout to nyx

nonnorm committed:
Fix syntax

nonnorm committed:
Rename nyxrc

nonnorm committed:
Get nyx to stop and start

nonnorm committed:
Add missing then

nonnorm committed:
Try again to get nyx to work

nonnorm committed:
Fix nameserver change

nonnorm committed:
Fix nyx scenes

nonnorm committed:
Add Tor relay to disabled scenes

nonnorm committed:
Make nyx update faster

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8385,6 +8416,223 @@ _EOF_
+
+			G_WHIP_BUTTON_CANCEL_TEXT='Skip'
+
+			if ! grep -q 'CONFIG_ENABLE_IPv6=1' /boot/dietpi.txt; then
```
The queck is done the opposite way, isn't it? Only ask to enter the IPv6 address if IPv6 is enabled:
```
if grep -q '^[[:blank:]]*CONFIG_ENABLE_IPv6=1' /boot/dietpi.txt; then
```
Leading `^[[:blank:]]*` just to be sure the line is not commented.

Btw, while most users will have IPv6 enabled, most won't have a static IPv6 address. Not sure how it works, if a domain name would be an alternative, but probably we could add the suggestion to leave it blank as well when no static public IPv6 address is available.

nonnorm committed:
Fix IPv6

MichaIng committed:
v6.34

+ DietPi-Software | Tor Relay: Fix and enhance whiptail format and loops

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8385,6 +8416,195 @@ _EOF_
+ExitPolicy accept *:64738
+ExitPolicy reject *:*
+_EOF_
+				G_CONFIG_INJECT 'nameserver[[:blank:]]' 'nameserver 127.0.0.1' /etc/resolv.conf
```
This may fail if there are multiple nameserver entries (`G_CONFIG_INJECT` requires the same setting to be present at most once) and it wont be persistent if `resolvconf` is installed, in which case `/etc/network/interfaces` needs to be altered to apply a persistent nameserver.

Currently there is no CLI to apply a DNS nameserver via DietPi tools,, but I'm working on one. Until then:
```
command -v resolvconf > /dev/null && G_EXEC sed -i 's/dns-nameservers.*$/dns-nameservers 127.0.0.1/' /etc/network/interfaces
G_EXEC eval "echo 'nameserver 127.0.0.1' > /etc/resolv.conf" # Unbound is running and configured in the very next config step, so it is safe to switch the DNS nameserver right now.
```

MichaIng committed:
v6.34

+ DietPi-CloudShell | Wait for Tor monitor process to be killed before continuing with next screne to avoid it printing over the next scene it case it hangs. As well kill it in case the "sleep" call fails, unlikely but possible.

MichaIng committed:
v6.34

+ DietPi-Services | Start Tor before WiFi hotspot just in case the latter uses the first (Tor hotspot)

MichaIng committed:
v6.34

+ DietPi-Software | Tor Relay: Minor whiptail and code format
+ DietPi-Software | Tor Relay: Apply local DNS resolver correctly based on resolvconf existance
+ DietPi-Software | Unbound: Restart service in every case at the end of config step.

MichaIng commented on dietpi/dietpi-cloudshell:
```diff
@@ -1066,6 +1066,9 @@
 	}
 
+	# Tor Relay: Include refresh rate delay and kill the monitor process afterwards
+	Update_Scene_9(){ nyx & sleep $REFRESH_RATE; killall -w nyx; }
```
Hmm, this is not yet optimal. When exiting (ctrl+c) DietPi-CloudShell while this scene is active, nyx stays running in background but printing to screen. This could be resolved by adding the `killall -w nyx` to the scripts exit trap.

Another thing: User might want to enable this as only scene, but it is killed and re-launched on refresh. Since it is not a static print but a self-refreshing monitor, it is a bid a special case. Probably we can move the kill to the start of the switcher loop (before or after the Update_Banner call): If the current scene is not the Tor monitor, kill nyx. Although not sure about this blank screen thing, needs to be reviewed/tested.
`nyx` must then only be called if it's not running yet. Since ps/pgrep methods are quite slow (recognisable delay), a flag variable would be best.

nonnorm replied:
Running `time pgrep "some process"` returns at most 0.03 seconds, on both my VM and RPi.

MichaIng replied:
Ah yeah sorry I forgot that this is a specific issue on Debian Bullseye:
```
2020-12-02 17:02:54 root@micha:/var/log# time pgrep mariadb
5162

real    0m0.322s
user    0m0.275s
sys     0m0.038s
```
I should report this to Debian bug tracker, as until Debian Buster it takes a fraction of that indeed. However, always good to avoid external binaries, where possible, especially when the positive result is required for an important step, like killing the background process here.

nonnorm replied:
`nyx` is working, but the graph glitches in the second run, sending blocks everywhere.

nonnorm committed:
Fix nyx refreshing

nonnorm committed:
Set NYX_ACTIVE variable

nonnorm committed:
Kill nyx on exit

nonnorm committed:
Only kil nyx if it is active

nonnorm committed:
Change exit trap to only kill nyx if it is running

nonnorm committed:
Switch to !=

nonnorm committed:
Add !

nonnorm committed:
Switch to ||

nonnorm committed:
Fix NYX_ACTIVE placement

nonnorm committed:
Fix syntax

nonnorm commented:
Everything works as far as I can tell.  The only thing I haven't tried is opening the ports up to the internet so people can actually connect.

MichaIng committed:
v6.34

+ DietPi-CloudShell | Declare NYX_ACTIVE once to assure it cannot be malformed from parent shell environment
+ DietPi-CloudShell | Use G_EXIT_CUSTOM for nyx kill exit trap to not override the G_INIT > G_EXIT of DietPi-Globals

MichaIng committed:
v6.34

+ DietPi-CloudShell | Pre-create nyx config directory before attempting to create config file
+ DietPi-CloudShell | Clear screen content before starting nyx
+ DietPi-CloudShell | Reset colour after intro animation

MichaIng committed:
v6.34

+ DietPi-Software | De-select conflicting Tor titles before Tor Hotspot would pull in WiFi Hotspot + simplify de-selection

MichaIng commented:
> The only thing I haven't tried is opening the ports up to the internet so people can actually connect.

Yes would be good if we could do that at latest during beta phase. However, from my end ready to merge now. I added some minor fixes, I hope you agree.

nonnorm commented:
They all look good to me.
